project/Student.cs

```diff
@@ -78,12 +82,57 @@ private float ValidateGPA(float gpa)
                 );
             }
 
-            if (gpa >= 3.0f && gpa <= 5.0f)
+            if (gpa is < 3.0f or > 5.0f)
             {
-                return gpa;
+                throw new ValidationException("GPA must be between 3.0 and 5.0.");
             }
 
-            throw new ValidationException("GPA must be between 3.0 and 5.0.");
+            return gpa;
+        }
+
+        private float CalculateGPA(List<Grade> grades)
+        {
+            uint sum = 0;
+            foreach (var grade in grades)
+            {
+                sum += grade.GradeValue;
+            }
+            return sum / grades.Count;
+        }
+
+        public void AddGrade(Grade grade)
+        {
+            // Does this make sense?
+            if (grade.Student == this)
```

Comment:
checks whether grade.Student reference is already set to the current student, looks good